hail/python/hail/ir/matrix_reader.py

```diff
@@ -20,18 +22,46 @@ def __eq__(self, other):
 
 
 class MatrixNativeReader(MatrixReader):
-    @typecheck_method(path=str)
-    def __init__(self, path):
+    @typecheck_method(path=str,
+                      intervals=nullable(sequenceof(anytype)),
+                      filter_intervals=bool)
+    def __init__(self, path, intervals, filter_intervals):
+        if intervals is not None:
+            t = hl.expr.impute_type(intervals)
```

Comment:
I think this logic should be in read_matrix_table -- we generally assume that parameters on the IR have been validated.

Reply:
also, shouldn't there be a check that the interval type agrees with the key type?

Reply:
Can't do that until this is wrapped in a `Table`/`MatrixTable` (python classes). We have no access to the key type at this point. I can certainly move much of the logic into `read(_matrix)_table`.

Reply:
Yeah, realized that after I commented and was looking at some of the code i master. I think it would be preferable to do the type lookup in read_matrix_table and then pass it into the IR node, but that's definitely out of scope of this change.